src/Sylius/Bundle/ThemeBundle/Templating/Locator/TemplateLocator.php

```diff
@@ -0,0 +1,270 @@
+     * @throws \InvalidArgumentException When the template file can not be found
+     */
+    public function locate($template, $currentPath = null, $first = true)
+    {
+        if (!$template instanceof TemplateReferenceInterface) {
+            throw new \InvalidArgumentException("The template must be an instance of TemplateReferenceInterface.");
+        }
+
+        $themes = $this->themeContext->getThemes();
+
+        $templatePath = $this->locateTemplateUsingThemes($template, $themes);
+
+        if (null === $templatePath) {
+            throw new \InvalidArgumentException(
+                sprintf('Unable to find template "%s" (using themes: %s).', $template, join(', ', $themes))
+            );
+        }
+
+        return $templatePath;
+    }
+
+    /**
+     * @param TemplateReferenceInterface $template
+     * @param ThemeInterface[] $themes
+     *
+     * @return null|string
+     */
+    protected function locateTemplateUsingThemes(TemplateReferenceInterface $template, array $themes)
+    {
+        foreach (array_merge($themes, [null]) as $theme) {
+            $result = $this->getCache($template, $theme);
+            if (null !== $result) {
+                return $result;
+            }
+        }
+
+        if (0 === strpos($template->getPath(), '@')) {
+            return $this->locateBundleTemplateUsingThemes($template, $themes);
+        } else {
+            return $this->locateAppTemplateUsingThemes($template, $themes);
+        }
+    }
+
+    /**
+     * @param TemplateReferenceInterface $template
+     * @param ThemeInterface[] $themes
+     *
+     * @return null|string
+     */
+    protected function locateBundleTemplateUsingThemes(TemplateReferenceInterface $template, array $themes)
+    {
+        $name = $template->getPath();
+
+        if (false !== strpos($name, '..')) {
+            throw new \RuntimeException(sprintf('File name "%s" contains invalid characters (..).', $name));
+        }
+
+        $bundleName = substr($name, 1);
+        $templatePath = '';
+        if (false !== strpos($bundleName, '/')) {
+            list($bundleName, $templatePath) = explode('/', $bundleName, 2);
+        }
+        if (0 !== strpos($templatePath, 'Resources')) {
+            throw new \RuntimeException('Template files have to be in Resources.');
+        }
+
+        $resourceBundle = null;
+        $bundles = $this->kernel->getBundle($bundleName, false);
+
+        $parameters = [
+            '%app_path%' => $this->appDir,
+            '%override_path%' => substr($templatePath, strlen('Resources/')),
+        ];
+
+        foreach ($this->paths['bundle_template'] as $path) {
+            foreach ($bundles as $bundle) {
+                $parameters = array_merge($parameters, [
+                    '%bundle_name%' => $bundle->getName(),
+                    '%bundle_path%' => $bundle->getPath(),
+                ]);
+
+                if (false === strpos($path, '%theme_path%')) {
```

Comment:
Maybe join this and nested condition in one? We could also (surprise :)) extract it to private method, named simple and nicely, and this allows us to avoid filthy ``else`` statement.